src/drivers/z3_driver.cpp

```diff
@@ -0,0 +1,153 @@
+#include "drivers/z3_driver.h"
+#include <c++/z3++.h>
+
+namespace expr {
+    struct z3_driver::impl {
+        z3::context c{};
+        z3::solver s;
+        auto as_symbol_value(const z3::expr& e) -> symbol_value_t;
+        auto as_z3_expression(const syntax_tree_t& tree) -> z3::expr;
+        auto as_z3_expression(const symbol_reference_t& ref) -> z3::expr;
+        auto as_z3_expression(const c_symbol_reference_t& ref) -> z3::expr;
+        auto as_z3_expression(const symbol_value_t& val) -> z3::expr;
+        void solve();
+        impl() : c{}, s{c} {}
+    };
+
+    z3_driver::z3_driver(const symbol_table_t& map) : environment{map}, pimpl{std::make_unique<z3_driver::impl>()}, driver{} {
+    }
+
+    z3_driver::~z3_driver() {}
+
+    int z3_driver::parse(const std::string &f) {
+        if (f.empty())
+            return 0;
+        file = f;
+        location.initialize(&file);
+        scan_begin();
+        yy::parser parse(this);
+        parse.set_debug_level(trace_parsing);
+        try {
+            int res = parse();
+            scan_end();
+            return res;
+        } catch (std::exception &e) {
+            error = e.what();
+            return 1;
+        }
+    }
+
+    auto z3_driver::get_symbol(const std::string &identifier) -> syntax_tree_t {
+#ifndef NDEBUG
+        if (!environment.contains(identifier))
+            throw std::out_of_range(identifier + " not found");
+#endif
+        return syntax_tree_t{environment.find(identifier)};
+    }
+
+    void z3_driver::add_tree(const syntax_tree_t& tree) {
+        pimpl->s.add(pimpl->as_z3_expression(tree)); // Note: Only accepts boolean expressions (will throw if not)
+        solve();
+        pimpl->s = z3::solver{pimpl->c};
+    }
+
+    void z3_driver::add_tree(const std::string& identifier, const syntax_tree_t& tree) {
+        std::cout << "z3_driver ignoring simple declaration" << std::endl;
+    }
+
+    void z3_driver::solve() {
+        switch (pimpl->s.check()) {
+            case z3::unsat:   std::cout << "unsat\n"; break;
+            case z3::unknown: std::cout << "unknown\n"; break;
```

Comment:
Maybe this should throw an exception instead of printing to cout - Look into that